ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/PullPhysicalPlanBuilder.java

```diff
@@ -0,0 +1,345 @@
+      if (currentLogicalNode instanceof ProjectNode) {
+        currentPhysicalOp = translateProjectNode((ProjectNode)currentLogicalNode);
+      } else if (currentLogicalNode instanceof FilterNode) {
+        currentPhysicalOp = translateFilterNode((FilterNode)currentLogicalNode);
+      } else if (currentLogicalNode instanceof DataSourceNode) {
+        currentPhysicalOp = translateDataSourceNode(
+            (DataSourceNode) currentLogicalNode, persistentQueryMetadata);
+        dataSourceOperator = (DataSourceOperator)currentPhysicalOp;
+      } else {
+        throw new KsqlException("Unrecognized logical node.");
+      }
+
+      if (prevPhysicalOp == null) {
+        rootPhysicalOp = currentPhysicalOp;
+      } else {
+        prevPhysicalOp.addChild(currentPhysicalOp);
+      }
+      prevPhysicalOp = currentPhysicalOp;
+      // For now assume only single source which is the case for pull queries
+      if (currentLogicalNode.getSources().isEmpty()) {
+        break;
+      }
+      currentLogicalNode = currentLogicalNode.getSources().get(0);
+    }
+
+    return new PullPhysicalPlan(
+        rootPhysicalOp,
+        ((ProjectOperator)rootPhysicalOp).getOutputSchema(),
+        queryId,
+        keys,
+        mat,
+        dataSourceOperator);
+  }
+
+  private ProjectOperator translateProjectNode(final ProjectNode logicalNode) {
+    LogicalSchema outputSchema = null;
+    boolean isStar = false;
+    if (isSelectStar(statement.getStatement().getSelect())) {
+      isStar = true;
+      outputSchema = buildSchema(mat.schema(), mat.windowType().isPresent());
+    } else {
+      final List<SelectExpression> projection = analysis.getSelectItems().stream()
+          .map(SingleColumn.class::cast)
+          .map(si -> SelectExpression
+              .of(si.getAlias().orElseThrow(IllegalStateException::new), si.getExpression()))
+          .collect(Collectors.toList());
+
+      outputSchema = selectOutputSchema(
+          executionContext, projection, mat.windowType());
+    }
+    return new ProjectOperator(
+      config,
+      metaStore,
+      mat,
+      analysis,
+      executionContext,
+      contextStacker,
+      logicalNode,
+      outputSchema,
+      isStar);
+  }
+
+  private SelectOperator translateFilterNode(final FilterNode logicalNode) {
+    whereInfo = WhereInfo.extractWhereInfo(analysis, persistentQueryMetadata);
+    return new SelectOperator(logicalNode);
+  }
+
+  private AbstractPhysicalOperator translateDataSourceNode(
+      final DataSourceNode logicalNode,
+      final PersistentQueryMetadata persistentQueryMetadata
+  ) {
+    if (whereInfo == null) {
+      throw new KsqlException("Pull queries must have a WHERE clause");
+    }
+    keys = whereInfo.getKeysBound().stream()
+        .map(keyBound -> asKeyStruct(keyBound, persistentQueryMetadata.getPhysicalSchema()))
+        .collect(ImmutableList.toImmutableList());
+
+    if (!whereInfo.getWindowBounds().isPresent()) {
```

Comment:
`getWindowBounds` is always present if it's windowed, regardless of what the bounds on the window are, right?  This is effectively an isWindowed check.

Reply:
If there are no window-bounds in the query, the query is not windowed. So, yes this is a `isWindowed` check